### ISSUE TITLE
Failing to build boost after update from 1.76 to 1.77

### ISSUE THREAD

**Issue by chausner** [microsoft/vcpkg#21286]
**Describe the bug**
After updating my local repo and rebuilding, I ran into a situation where I couldn't build boost packages anymore because of a conflict with older boost versions that were still installed for a different triplet:
```
Starting package 1/47: boost-exception:x86-windows
Building package boost-exception[core]:x86-windows...
-- Using cached boostorg-exception-boost-1.77.0.tar.gz.
-- Cleaning sources at D:/vcpkg/buildtrees/boost-exception/src/ost-1.77.0-6531b39968.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source D:/vcpkg/downloads/boostorg-exception-boost-1.77.0.tar.gz
-- Using source at D:/vcpkg/buildtrees/boost-exception/src/ost-1.77.0-6531b39968.clean
-- Building release...
CMake Error at scripts/cmake/vcpkg_configure_cmake.cmake:110 (message):
  The boost-exception port already depends on vcpkg-cmake; using both
  vcpkg-cmake and vcpkg_configure_cmake in the same port is unsupported.
Call Stack (most recent call first):
  installed/x64-windows/share/boost-build/boost-modular-build.cmake:88 (vcpkg_configure_cmake)
  installed/x64-windows/share/boost-build/boost-modular-build.cmake:113 (unix_build)
  ports/boost-exception/portfile.cmake:15 (boost_modular_build)
  scripts/ports.cmake:142 (include)


Error: Building package boost-exception:x86-windows failed with: BUILD_FAILED
```

The bug appears to occur because 
* boost 1.76 was still using vcpkg_configure_cmake while boost 1.77 uses vcpkg_cmake_configure
* Even though I was building for x86-windows, vcpkg for some reason included `boost-build/boost-modular-build.cmake` from the x64-windows install (visible from the log above) which is still boost 1.76. Thus a newer CMake script using vcpkg-cmake included an older CMake script which still used vcpkg_configure_cmake.

**Environment**
- OS: Windows
- Compiler: MSVC

**To Reproduce**
Steps to reproduce the behavior (I haven't validated the steps but I would expect these to reproduce the issue in theory):
1. Checkout a state of the vcpkg repo with boost version 1.76, i.e. before it was updated to 1.77.
2. vcpkg install boost:x64-windows ; vcpkg remove boost-exception:x64-windows
3. Update the repository to ce73eebf08f4033b16d390ed75f15c58cd84cdd4, a state where boost has been updated to 1.77
4. vcpkg install boost:x86-windows

After running `vcpkg remove boost-uninstall:x64-windows --recurse` to fully remove the older 1.76 installation for x64-windows, I was able to build boost:x86-windows successfully.

**Expected behavior**
I expect boost 1.77 to be installed successfully for x86-windows, while boost 1.76 should remain installed for x64-windows.

**Failure logs** 
See log above.

**Comment by autoantwort** [comment]
> `vcpkg for some reason included boost-build/boost-modular-build.cmake from the x64-windows` 

Because x64-windows ist your host triplet, and boost-build is build for the host triplet.  
To fix your problem you have to pass `--host-triplet=x86-windows` to vcpkg

**Comment by chausner** [comment]
@autoantwort So you would not consider this a bug in vcpkg or the ports at all? If this is expected behavior, I still think there could be better diagnostics or error messages at least. It took me quite a while to figure out what the problem was and how to workaround it. Also I do not quite understand yet why I would need to pass `--host-triplet=x86-windows` here if it was never needed for any other packages I installed in the past.

**Comment by autoantwort** [comment]
Yeah I think this is expected behavior. 

> I still think there could be better diagnostics 

The "problem" is that your error can only happen with boost and only if you use the classic mode and you install boost for a non host triplet >= 1.77 and have an existing installation for the host triplet with version < 1.77. Otherwise you can't get the error. 

> Also I do not quite understand yet why I would need to pass --host-triplet=x86-windows here if it was never needed for any other packages I installed in the past.

There are two ways for cmake builds (`vcpkg_configure_cmake`(deprecated) and `vcpkg_cmake_configure`) which are incompatible and you can't use them at the same time. boost switched to the new version in 1.77. In your situation boost-build for the host-triplet is in version 1.76 while your packages are in version 1.77 and then you get the error that you can't use boost cmake configure things at the same time. You can update boost for the host-triplet to 1.77 or you change the host triplet to 1.76

**Comment by chausner** [comment]
@autoantwort Thank you for your explanation. Good to know this is an issue that is specific to the boost transition from 1.76 to 1.77 and should not happen with other ports. I opened this issue primarily because I thought it was a bug that vcpkg devs may want to fix.

If it is not considered a bug, feel free to close this issue then. I was able to find a workaround myself already, and @autoantwort  pointed out about `--host-triplet`, so I am not blocked by this issue anymore myself.
